### PR TITLE
Add feature to map pairwise distances matrix to disk while computing the cluster matrix in ASSET

### PR DESCRIPTION
**Pull request by kohlerca**
The `cluster_matrix_entries` step in ASSET uses DBSCAN on the generated mask matrix, to extract the clusters associated with the SSEs being identified. As a custom distance metric was used, the implementation in ASSET first precomputes all pairwise distances and passes them to the `sklearn` function that performs DBSCAN.

Depending on the number of significant points being clustered, the memory requirements for this distance matrix increase. An option to compute the distances in chunks is available when memory errors occurred, but this still requires the allocation of an NxN matrix with `float32` precision in memory (where N is the number of significant points found in the mask matrix). As the data length increases, it is expected that the number of significant points will also increase, and the computation would be unfeasible as this matrix can quickly reach TiB sizes.

This PR adds an option to pass a location for a file on the disk, that will be used to map the pairwise distance matrix to a file. Therefore, no allocation in memory is needed, but the operations will be slower. This was implemented to be used only if already using chunking (`working_memory` parameter). As additional options, the user can produce verbose messages for debugging, and also has the option of keeping the file (which could be useful if the user wants to do some analyses on the distances, such as k-distance graphs).

### CODE REVIEW

**Review by Moritz-Alexander-Kern — CHANGES_REQUESTED** [review]
Nice functionality, thanks for this enhancement.

**Comment by Moritz-Alexander-Kern** [comment]
Consider lines:

https://github.com/NeuralEnsemble/elephant/blob/646cbd444367f2b8918415536d7fe87948d745bb/elephant/asset/asset.py#L416-L432

Run the following example:
```python
import numpy as np
import elephant.asset.asset as asset
stretchedmetric2d = asset._stretched_metric_2d

def my_func():
    nr_points = 10000

    x = np.arange(nr_points)
    D = stretchedmetric2d(x, x, stretch=1, ref_angle=45)

if __name__ == '__main__':
    my_func()

```
With memory_profiler:

```python
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   445    108.6 MiB      0.0 MiB           1       if working_memory is None:
   446                                                 # Compute the matrix D[i, j] of euclidean distances among points
   447                                                 # i and j
   448    872.7 MiB    764.1 MiB           1           D = pairwise_distances(points)
   449                                         
   450                                                 # Compute the angular coefficients of the line between each pair of
   451                                                 # points
   452                                         
   453                                                 # dX[i,j]: x difference between points i and j
   454                                                 # dY[i,j]: y difference between points i and j
   455   1635.7 MiB    763.0 MiB           1           dX = x_array.T - x_array
   456   2398.6 MiB    762.9 MiB           1           dY = y_array.T - y_array
   457                                         
   458                                                 # Compute the matrix Theta of angles between each pair of points
   459   3161.7 MiB    763.1 MiB           1           theta = np.arctan2(dY, dX, dtype=np.float64)
   460                                         
   461   3161.7 MiB      0.0 MiB           1           stretch_mat = calculate_stretch_mat(theta, D)
...
   548                                             # Return the stretched distance matrix
   549   3161.7 MiB      0.0 MiB           1       return stretch_mat
```

Change Lines 416-432 to:
```python

    if working_memory is None:
        # Compute the angular coefficients of the line between each pair of
        # points

        # calculate difference between all pairs of abscissas and ordinates
        d_array = lambda array: array.T - array

        # Compute the matrix Theta of angles between each pair of points
        theta = lambda y_array, x_array: np.arctan2(d_array(y_array),
                                                    d_array(x_array),
                                                    dtype=np.float64)
        # Compute the matrix D[i, j] of euclidean distances among points
        # i and j
        D = pairwise_distances
        stretch_mat = calculate_stretch_mat(theta(y_array, x_array), D(points))
```
With memory_profiler:

```python
   444    109.1 MiB      0.0 MiB           1       if working_memory is None:
   445                                                 # Compute the angular coefficients of the line between each pair of
   446                                                 # points
   447                                         
   448                                                 # calculate difference between all pairs of abscissas and ordinates
   449   1635.1 MiB   1526.0 MiB           5           d_array = lambda array: array.T - array
   450                                         
   451                                                 # Compute the matrix Theta of angles between each pair of points
   452    872.5 MiB   -762.7 MiB           4           theta = lambda y_array, x_array: np.arctan2(d_array(y_array),
   453   1635.1 MiB      0.0 MiB           1                                                       d_array(x_array),
   454   1635.1 MiB      0.0 MiB           1                                                       dtype=np.float64)
   455                                                 # Compute the matrix D[i, j] of euclidean distances among points
   456                                                 # i and j
   457    109.1 MiB      0.0 MiB           1           D = pairwise_distances
   458    873.5 MiB   -762.9 MiB           1           stretch_mat = calculate_stretch_mat(theta(y_array, x_array), D(points))
...
   546    873.5 MiB      0.0 MiB           1       return stretch_mat

```

What is your opinion on this?

File: elephant/asset/asset.py [hunk 1]
```diff
@@ -379,12 +381,37 @@ def _stretched_metric_2d(x, y, stretch, ref_angle, working_memory=None):
...
         Reference angle in degrees (i.e., the inclination along which the
         stretching factor is 1).
+    working_memory : int, optional
+        The sought maximum memory in MiB for temporary distance matrix chunks.
+        When None (default), no chunking is performed. This parameter is passed
+        directly to `sklearn.metrics.pairwise_distances_chunked` function and
+        it has no influence on the outcome matrix. Instead, it control the
+        memory VS speed trade-off.
+        Default: None
+    disk_array : file-like, optional
```
[the remark this answers is not in the document]
Comment (Moritz-Alexander-Kern) [reply elephant/asset/asset.py:391]: mapped_array_file

File: elephant/asset/asset.py [hunk 2]
```diff
@@ -432,20 +459,88 @@ def calculate_stretch_mat(theta_mat, D_mat):
...
+        # Size is computed for a float32 matrix. The function
+        # `pairwise_distances_chunked` returns half of the possible size.
+        estimated_chunk = max(
+            ((working_memory * 1024 * 1024) // (len(y) * 4)) // 2, 1)
+        
+        # The number of rows in a chunk cannot be larger than the maximum
+        estimated_chunk = min(len(x), estimated_chunk)
+
+        # Compute the number of iterations needed
+        it_todo = len(x) // estimated_chunk
```
[the remark this answers is not in the document]
Comment (Moritz-Alexander-Kern) [reply elephant/asset/asset.py:474]: it_todo

**Review by Moritz-Alexander-Kern — APPROVED** [review]
